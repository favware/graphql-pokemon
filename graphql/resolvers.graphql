type Query {
  "Gets the details on a Pokémon ability, using the ability name"
  getAbility("The ability to look up" ability: AbilitiesEnum!): Ability!

  """
  Gets details on a Pokémon ability, using a fuzzy search on name

  This can be used to find multiple results based on the query

  You can provide `take` to limit the amount of abilities to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.
  """
  getFuzzyAbility(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The ability to fuzzily search"
    ability: String!
  ): [Ability!]!

  """
  Gets details on a single Pokémon based on National Pokédex number

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getPokemonByDexNumber(
    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true

    "The International PokéDex number of the Pokémon to look up"
    number: Int!
  ): Pokemon!

  """
  Gets details on a single Pokémon based on species name

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getPokemon(
    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true

    "The Pokémon to look up"
    pokemon: PokemonEnum!
  ): Pokemon!

  """
  Gets details on one or more Pokémon based on species name

  You can provide `take` to limit the amount of Pokémon to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getFuzzyPokemon(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The Pokémon to fuzzily search"
    pokemon: String!

    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true
  ): [Pokemon!]!

  """
  Returns a list of all the known Pokémon.

  For every Pokémon all the data on each requested field is returned.

  **_NOTE:_ To skip all CAP Pokémon, PokéStar Pokémon, and Missingno provide an `offset` of 87**

  You can provide `take` to limit the amount of Pokémon to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  While the API will currently not rate limit the usage of this query, it may do so in the future.

  It is advisable to cache responses of this query.
  """
  getAllPokemon(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1273

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true
  ): [Pokemon!]!

  "Gets the details on a Pokémon item, using the item name"
  getItem("The item to look up" item: ItemsEnum!): Item!

  """
  Gets details on a Pokémon item, using a fuzzy search on name

  This can be used to find multiple results based on the query

  By default only 1 result is returned. You can provide the arguments `take`, `offset`, and `reverse` to modify this behaviour.
  """
  getFuzzyItem(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The item to fuzzily search"
    item: String!
  ): [Item!]!

  """
  Gets the learnsets for a given Pokémon and move.

  Multiple moves are possible by putting them in an array: `[move1, move2]`.

  You can also apply a generation filter (only results for the given generation will be returned) with the generation argument
  """
  getLearnset(
    "The generation filter to apply"
    generation: Int

    "The moves to match against the Pokémon"
    moves: [MovesEnum!]!

    "The Pokémon for which to get the learnset"
    pokemon: PokemonEnum!
  ): Learnset!

  """
  Gets the learnset for a given Pokémon and move.

  A fuzzy search is performed to find a matching Pokémon and move

  Multiple moves are possible by putting them in an array: `[move1, move2]`.

  You can also apply a generation filter (only results for the given generation will be returned) with the generation argument
  """
  getFuzzyLearnset(
    "The generation filter to apply"
    generation: Int

    "The moves to match against the Pokémon"
    moves: [String!]!

    "The Pokémon for which to get the learnset"
    pokemon: String!
  ): Learnset!

  "Gets the details on a Pokémon move, using the move name"
  getMove("The move to look up" move: MovesEnum!): Move!

  """
  Gets details on a Pokémon move, using a fuzzy search on name

  This can be used to find multiple results based on the query

  By default only 1 result is returned. You can provide the arguments `take`, "offset", and "reverse" to modify this behaviour.
  """
  getFuzzyMove(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The move to fuzzily search"
    move: String!
  ): [Move!]!

  "Gets the type matchup data for the given type or types"
  getTypeMatchup("The primary type to check" primaryType: TypesEnum!, "The secondary type to check" secondaryType: TypesEnum): TypeMatchup!
}
