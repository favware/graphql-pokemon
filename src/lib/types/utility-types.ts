/**
 * Removes `null`, `undefined` and `?` from an object
 * This essentially combines `Required<T>`,
 * `NonNullable<T>` and `NonUndefined<T>`
 */
export type NonNullish<T> = {
  [P in keyof T]-?: Exclude<T[P], undefined | null>;
};

export type UnwrapArray<T extends unknown[]> = T extends (infer U)[] ? U : never;

export type Generation = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
