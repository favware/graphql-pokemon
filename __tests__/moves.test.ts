import { GraphQLError } from 'graphql';
import { formatResponse, gCall, gql } from './testUtils/testUtils';
import { DataResponse } from './testUtils/types';

describe('getMoveDetailsByName', () => {
  const getMoveDetailsByName = gql`
    query($move: Moves!) {
      getMoveDetailsByName(move: $move) {
        name
      }
    }
  `;

  test('GIVEN a valid move THEN returns MoveEntry', async () => {
    const { data } = (await gCall({
      source: getMoveDetailsByName,
      variableValues: { move: 'dragondance' }
    }).then(formatResponse)) as DataResponse<'getMoveDetailsByName'>;

    expect(data.getMoveDetailsByName.name).toBe('Dragon Dance');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move THEN returns error', async () => {
    const data = await gCall({
      source: getMoveDetailsByName,
      variableValues: { move: 'totally_invalid_move' }
    });

    data.errors!.forEach((error) => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});

describe('getMoveDetailsByFuzzy', () => {
  const getMoveDetailsByFuzzy = gql`
    query($move: String!, $skip: Int, $take: Int, $reverse: Boolean) {
      getMoveDetailsByFuzzy(move: $move, skip: $skip, take: $take, reverse: $reverse) {
        name
      }
    }
  `;

  test('GIVEN a valid move THEN returns MoveEntry', async () => {
    const { data } = (await gCall({
      source: getMoveDetailsByFuzzy,
      variableValues: { move: 'dragondance' }
    }).then(formatResponse)) as DataResponse<'getMoveDetailsByFuzzy'>;

    expect(data.getMoveDetailsByFuzzy.name).toBe('Dragon Dance');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy move THEN returns MoveEntry', async () => {
    const { data } = (await gCall({
      source: getMoveDetailsByFuzzy,
      variableValues: { move: 'dragon' }
    }).then(formatResponse)) as DataResponse<'getMoveDetailsByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy move and pagination THEN returns MoveEntry', async () => {
    const { data } = (await gCall({
      source: getMoveDetailsByFuzzy,
      variableValues: {
        move: 'dragon',
        skip: 2,
        take: 1,
        reverse: true
      }
    }).then(formatResponse)) as DataResponse<'getMoveDetailsByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move THEN returns error', async () => {
    const data = await gCall({
      source: getMoveDetailsByFuzzy,
      variableValues: { move: 'totally_invalid_move' }
    });

    data.errors!.forEach((error) => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move and pagination THEN returns error', async () => {
    const data = await gCall({
      source: getMoveDetailsByFuzzy,
      variableValues: {
        move: 'totally_invalid_move',
        skip: 0,
        take: 1,
        reverse: false
      }
    });

    data.errors!.forEach((error) => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});

describe('getMoveByFuzzy', () => {
  const getMoveByFuzzy = gql`
    query($move: String!, $skip: Int, $take: Int, $reverse: Boolean) {
      getMoveByFuzzy(move: $move, skip: $skip, take: $take, reverse: $reverse)
    }
  `;

  test('GIVEN a valid move THEN returns JSONObject', async () => {
    const { data } = (await gCall({
      source: getMoveByFuzzy,
      variableValues: { move: 'dragondance' }
    }).then(formatResponse)) as DataResponse<'getMoveByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy move THEN returns JSONObject', async () => {
    const { data } = (await gCall({
      source: getMoveByFuzzy,
      variableValues: { move: 'dragon' }
    }).then(formatResponse)) as DataResponse<'getMoveByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy move and pagination THEN returns MoveEntry', async () => {
    const { data } = (await gCall({
      source: getMoveByFuzzy,
      variableValues: {
        move: 'dragon',
        skip: 2,
        take: 1,
        reverse: true
      }
    }).then(formatResponse)) as DataResponse<'getMoveByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move THEN returns empty array', async () => {
    const { data } = (await gCall({
      source: getMoveByFuzzy,
      variableValues: { move: 'totally_invalid_move' }
    }).then(formatResponse)) as DataResponse<'getMoveByFuzzy'>;

    expect(data.getMoveByFuzzy).toStrictEqual([]);
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move and pagination THEN returns empty array', async () => {
    const { data } = (await gCall({
      source: getMoveByFuzzy,
      variableValues: {
        move: 'totally_invalid_move',
        skip: 0,
        take: 1,
        reverse: false
      }
    }).then(formatResponse)) as DataResponse<'getMoveByFuzzy'>;

    expect(data.getMoveByFuzzy).toStrictEqual([]);
    expect(data).toMatchSnapshot();
  });
});

describe('getMoveByName', () => {
  const getMoveByName = gql`
    query($move: Moves!) {
      getMoveByName(move: $move)
    }
  `;

  test('GIVEN a valid move THEN returns JSONObject', async () => {
    const { data } = (await gCall({
      source: getMoveByName,
      variableValues: { move: 'dragondance' }
    }).then(formatResponse)) as DataResponse<'getMoveByName'>;

    expect(data.getMoveByName.name).toBe('Dragon Dance');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid move THEN returns error', async () => {
    const data = await gCall({
      source: getMoveByName,
      variableValues: { move: 'totally_invalid_move' }
    });

    data.errors!.forEach((error) => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});
