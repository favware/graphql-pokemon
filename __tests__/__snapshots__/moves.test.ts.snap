// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMoveByFuzzy GIVEN a fuzzy move THEN returns JSONObject 1`] = `
Object {
  "getMoveByFuzzy": Array [
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "120",
        "category": "Physical",
        "contestType": "Beautiful",
        "desc": "Lowers the user's Defense and Special Defense by 1 stage.",
        "maxMovePower": 140,
        "name": "Dragon Ascent",
        "pp": 5,
        "priority": 0,
        "shortDesc": "Lowers the user's Defense and Sp. Def by 1.",
        "target": "Any",
        "type": "Flying",
      },
      "refIndex": 162,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "60",
        "category": "Special",
        "contestType": "Cool",
        "desc": "Has a 30% chance to paralyze the target.",
        "maxMovePower": 110,
        "name": "Dragon Breath",
        "pp": 20,
        "priority": 0,
        "shortDesc": "30% chance to paralyze the target.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 163,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "80",
        "category": "Physical",
        "contestType": "Cool",
        "desc": "No additional effect.",
        "maxMovePower": 130,
        "name": "Dragon Claw",
        "pp": 15,
        "priority": 0,
        "shortDesc": "No additional effect.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 164,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "0",
        "category": "Status",
        "contestType": "Cool",
        "desc": "Raises the user's Attack and Speed by 1 stage.",
        "name": "Dragon Dance",
        "pp": 20,
        "priority": 0,
        "shortDesc": "Raises the user's Attack and Speed by 1.",
        "target": "Self",
        "type": "Dragon",
      },
      "refIndex": 165,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "50",
        "category": "Physical",
        "desc": "Hits twice. If the first hit breaks the target's substitute, it will take damage for the second hit. In Double Battles, this move attempts to hit the targeted Pokémon and its ally once each. If hitting one of these Pokémon would be prevented by immunity, protection, semi-invulnerability, an Ability, or accuracy, it attempts to hit the other Pokémon twice instead. If this move is redirected, it hits that target twice.",
        "maxMovePower": 130,
        "name": "Dragon Darts",
        "pp": 10,
        "priority": 0,
        "shortDesc": "Hits twice. Doubles: Tries to hit each foe once.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 166,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "150",
        "category": "Special",
        "desc": "Power is equal to (user's current HP * 150 / user's maximum HP), rounded down, but not less than 1.",
        "name": "Dragon Energy",
        "pp": 5,
        "priority": 0,
        "shortDesc": "Less power as user's HP decreases. Hits foe(s).",
        "target": "Adjacent Foes",
        "type": "Dragon",
      },
      "refIndex": 167,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "90",
        "category": "Physical",
        "contestType": "Tough",
        "desc": "No additional effect.",
        "maxMovePower": 130,
        "name": "Dragon Hammer",
        "pp": 15,
        "priority": 0,
        "shortDesc": "No additional effect.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 168,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "85",
        "category": "Special",
        "contestType": "Beautiful",
        "desc": "No additional effect.",
        "maxMovePower": 130,
        "name": "Dragon Pulse",
        "pp": 10,
        "priority": 0,
        "shortDesc": "No additional effect.",
        "target": "Any",
        "type": "Dragon",
      },
      "refIndex": 169,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "40",
        "category": "Special",
        "contestType": "Cool",
        "desc": "Deals 40 HP of damage to the target.",
        "isNonstandard": "Past",
        "name": "Dragon Rage",
        "pp": 10,
        "priority": 0,
        "shortDesc": "Always does 40 HP of damage.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 170,
    },
    Object {
      "item": Object {
        "accuracy": 75,
        "basePower": "100",
        "category": "Physical",
        "contestType": "Tough",
        "desc": "Has a 20% chance to make the target flinch. Damage doubles and no accuracy check is done if the target has used Minimize while active.",
        "maxMovePower": 130,
        "name": "Dragon Rush",
        "pp": 10,
        "priority": 0,
        "shortDesc": "20% chance to make the target flinch.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 171,
    },
  ],
}
`;

exports[`getMoveByFuzzy GIVEN a fuzzy move and pagination THEN returns MoveEntry 1`] = `
Object {
  "getMoveByFuzzy": Array [
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "40",
        "category": "Special",
        "contestType": "Cool",
        "desc": "Deals 40 HP of damage to the target.",
        "isNonstandard": "Past",
        "name": "Dragon Rage",
        "pp": 10,
        "priority": 0,
        "shortDesc": "Always does 40 HP of damage.",
        "target": "Normal",
        "type": "Dragon",
      },
      "refIndex": 170,
    },
  ],
}
`;

exports[`getMoveByFuzzy GIVEN a valid move THEN returns JSONObject 1`] = `
Object {
  "getMoveByFuzzy": Array [
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "0",
        "category": "Status",
        "contestType": "Cool",
        "desc": "Raises the user's Attack and Speed by 1 stage.",
        "name": "Dragon Dance",
        "pp": 20,
        "priority": 0,
        "shortDesc": "Raises the user's Attack and Speed by 1.",
        "target": "Self",
        "type": "Dragon",
      },
      "refIndex": 165,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "120",
        "category": "Physical",
        "contestType": "Beautiful",
        "desc": "Lowers the user's Defense and Special Defense by 1 stage.",
        "maxMovePower": 140,
        "name": "Dragon Ascent",
        "pp": 5,
        "priority": 0,
        "shortDesc": "Lowers the user's Defense and Sp. Def by 1.",
        "target": "Any",
        "type": "Flying",
      },
      "refIndex": 162,
    },
    Object {
      "item": Object {
        "accuracy": 100,
        "basePower": "150",
        "category": "Special",
        "desc": "Power is equal to (user's current HP * 150 / user's maximum HP), rounded down, but not less than 1.",
        "name": "Dragon Energy",
        "pp": 5,
        "priority": 0,
        "shortDesc": "Less power as user's HP decreases. Hits foe(s).",
        "target": "Adjacent Foes",
        "type": "Dragon",
      },
      "refIndex": 167,
    },
  ],
}
`;

exports[`getMoveByFuzzy GIVEN an invalid move THEN returns empty array 1`] = `
Object {
  "getMoveByFuzzy": Array [],
}
`;

exports[`getMoveByFuzzy GIVEN an invalid move and pagination THEN returns empty array 1`] = `
Object {
  "getMoveByFuzzy": Array [],
}
`;

exports[`getMoveByName GIVEN a valid move THEN returns JSONObject 1`] = `
Object {
  "getMoveByName": Object {
    "accuracy": 100,
    "basePower": "0",
    "category": "Status",
    "contestType": "Cool",
    "desc": "Raises the user's Attack and Speed by 1 stage.",
    "name": "Dragon Dance",
    "pp": 20,
    "priority": 0,
    "shortDesc": "Raises the user's Attack and Speed by 1.",
    "target": "Self",
    "type": "Dragon",
  },
}
`;

exports[`getMoveByName GIVEN an invalid move THEN returns error 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$move" got invalid value "totally_invalid_move"; Value "totally_invalid_move" does not exist in "Moves" enum.],
  ],
}
`;

exports[`getMoveDetailsByFuzzy GIVEN a fuzzy move THEN returns MoveEntry 1`] = `
Object {
  "getMoveDetailsByFuzzy": Object {
    "name": "Dragon Ascent",
  },
}
`;

exports[`getMoveDetailsByFuzzy GIVEN a fuzzy move and pagination THEN returns MoveEntry 1`] = `
Object {
  "getMoveDetailsByFuzzy": Object {
    "name": "Dragon Rage",
  },
}
`;

exports[`getMoveDetailsByFuzzy GIVEN a valid move THEN returns MoveEntry 1`] = `
Object {
  "getMoveDetailsByFuzzy": Object {
    "name": "Dragon Dance",
  },
}
`;

exports[`getMoveDetailsByFuzzy GIVEN an invalid move THEN returns error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Failed to get data for move: totally_invalid_move],
  ],
}
`;

exports[`getMoveDetailsByFuzzy GIVEN an invalid move and pagination THEN returns error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Failed to get data for move: totally_invalid_move],
  ],
}
`;

exports[`getMoveDetailsByName GIVEN a valid move THEN returns MoveEntry 1`] = `
Object {
  "getMoveDetailsByName": Object {
    "name": "Dragon Dance",
  },
}
`;

exports[`getMoveDetailsByName GIVEN an invalid move THEN returns error 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$move" got invalid value "totally_invalid_move"; Value "totally_invalid_move" does not exist in "Moves" enum.],
  ],
}
`;
